File: plugin/jvm/src/main/java/com/twosigma/beaker/table/TableDisplay.java
/*
 *  Copyright 2014 TWO SIGMA OPEN SOURCE, LLC
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.twosigma.beaker.table;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.concurrent.TimeUnit;
import java.util.logging.Logger;
import java.util.Set;

import com.twosigma.beaker.jvm.serialization.BasicObjectSerializer;

import com.twosigma.beaker.jvm.serialization.BeakerObjectConverter;
import com.twosigma.beaker.table.format.TableDisplayStringFormat;
import com.twosigma.beaker.table.format.ValueStringFormat;
import com.twosigma.beaker.table.highlight.TableDisplayCellHighlighter;
import com.twosigma.beaker.table.renderer.TableDisplayCellRenderer;

public class TableDisplay extends ObservableTableDisplay {
  public static final String TABLE_DISPLAY_SUBTYPE = "TableDisplay";
  public static final String LIST_OF_MAPS_SUBTYPE = "ListOfMaps";
  public static final String MATRIX_SUBTYPE = "Matrix";
  public static final String DICTIONARY_SUBTYPE = "Dictionary";
  private final static Logger logger = Logger.getLogger(TableDisplay.class.getName());
  private final List<List<?>> values;
  private final List<String> columns;
  private final List<String> classes;
  private final String subtype;

  private TimeUnit stringFormatForTimes;
  private Map<ColumnType, TableDisplayStringFormat> stringFormatForType = new HashMap<>();
  private Map<String, TableDisplayStringFormat> stringFormatForColumn = new HashMap<>();
  private Map<ColumnType, TableDisplayCellRenderer> rendererForType = new HashMap<>();
  private Map<String, TableDisplayCellRenderer> rendererForColumn = new HashMap<>();
  private Map<ColumnType, TableDisplayAlignmentProvider> alignmentForType = new HashMap<>();
  private Map<String, TableDisplayAlignmentProvider> alignmentForColumn = new HashMap<>();
  private Map<String, Boolean> columnsFrozen = new HashMap<>();
  private Map<String, Boolean> columnsFrozenRight = new HashMap<>();
  private Map<String, Boolean> columnsVisible = new HashMap<>();
  private List<String> columnOrder = new ArrayList<>();
  private List<TableDisplayCellHighlighter> cellHighlighters = new ArrayList<>();
  private List<List<String>> tooltips = new ArrayList<>();

  public TableDisplay(List<List<?>> v, List<String> co, List<String> cl) {
    values = v;
    columns = co;
    classes = cl;
    subtype = TABLE_DISPLAY_SUBTYPE;
  }

  public TableDisplay(Collection<Map<?,?>> v) {
    this(v, new BasicObjectSerializer());
  }

  public TableDisplay(Collection<Map<?,?>> v, BeakerObjectConverter serializer) {
    values = new ArrayList<List<?>>();
    columns = new ArrayList<String>();
    classes = new ArrayList<String>();
    subtype = LIST_OF_MAPS_SUBTYPE;

    // create columns
    for(Map<?,?> m : v) {
      Set<?> w = m.entrySet();
      for (Object s : w) {
        Entry<?,?> e = (Entry<?, ?>) s;
        String c = e.getKey().toString();
        if (!columns.contains(c)) {
          columns.add(c);
          String n = e.getValue()!=null ? e.getValue().getClass().getName() : "string";
          classes.add(serializer.convertType(n));
        }
      }
    }

    // now build values
    for(Map<?,?> m : v) {
      List<Object> vals = new ArrayList<Object>();
      for (String cn : columns) {
        if (m.containsKey(cn)){
          vals.add(getValueForSerializer( m.get(cn), serializer));
        }
        else
          vals.add(null);
      }
      values.add(vals);
    }
  }

  public TimeUnit getStringFormatForTimes() {
    return stringFormatForTimes;
  }

  public void setStringFormatForTimes(TimeUnit stringFormatForTimes) {
    this.stringFormatForTimes = stringFormatForTimes;
  }

  public Map<ColumnType, TableDisplayStringFormat> getStringFormatForType() {
    return stringFormatForType;
  }

  public void setStringFormatForType(ColumnType type, TableDisplayStringFormat format) {
    this.stringFormatForType.put(type, format);
  }

  public Map<String, TableDisplayStringFormat> getStringFormatForColumn() {
    return stringFormatForColumn;
  }

  public void setStringFormatForColumn(String column, TableDisplayStringFormat format) {
    this.stringFormatForColumn.put(column, format);
  }

  public void setStringFormatForColumn(String column, Object closure) {
    int colIndex = this.columns.indexOf(column);
    if (colIndex == -1) {
      throw new IllegalArgumentException("Column " + column + " doesn't exist");
    }
    List<String> formattedValues = new ArrayList<>();
    try {
      for (int row = 0; row < this.values.size(); row++) {
        Object value = this.values.get(row).get(colIndex);
        Object[] params = new Object[]{ value, row, column, this };
        formattedValues.add((String) runClosure(closure, params));
      }
    } catch (Throwable e) {
      throw new IllegalArgumentException("Can not create format using closure.", e);
    }
    this.stringFormatForColumn.put(column, new ValueStringFormat(column, formattedValues));
  }

  public Map<ColumnType, TableDisplayCellRenderer> getRendererForType() {
    return rendererForType;
  }

  public void setRendererForType(ColumnType type, TableDisplayCellRenderer renderer) {
    this.rendererForType.put(type, renderer);
  }

  public Map<String, TableDisplayCellRenderer> getRendererForColumn() {
    return rendererForColumn;
  }

  public void setRendererForColumn(String column, TableDisplayCellRenderer renderer) {
    this.rendererForColumn.put(column, renderer);
  }

  public Map<ColumnType, TableDisplayAlignmentProvider> getAlignmentForType() {
    return alignmentForType;
  }

  public void setAlignmentProviderForType(ColumnType type, TableDisplayAlignmentProvider alignmentProvider) {
    this.alignmentForType.put(type, alignmentProvider);
  }

  public Map<String, TableDisplayAlignmentProvider> getAlignmentForColumn() {
    return alignmentForColumn;
  }

  public void setAlignmentProviderForColumn(String column, TableDisplayAlignmentProvider alignmentProvider) {
    this.alignmentForColumn.put(column, alignmentProvider);
  }

  public Map<String, Boolean> getColumnsFrozen() {
    return columnsFrozen;
  }

  public void setColumnFrozen(String column, boolean frozen) {
    this.columnsFrozen.put(column, frozen);
  }

  public Map<String, Boolean> getColumnsFrozenRight() {
    return columnsFrozenRight;
  }

  public void setColumnFrozenRight(String column, boolean frozen) {
    this.columnsFrozenRight.put(column, frozen);
  }

  public Map<String, Boolean> getColumnsVisible() {
    return columnsVisible;
  }

  public void setColumnVisible(String column, boolean visible) {
    this.columnsVisible.put(column, visible);
  }

  public List<String> getColumnOrder() {
    return columnOrder;
  }

  public List<TableDisplayCellHighlighter> getCellHighlighters() {
    return cellHighlighters;
  }

  public void addCellHighlighter(TableDisplayCellHighlighter cellHighlighter) {
    this.cellHighlighters.add(cellHighlighter);
  }

  public void removeAllCellHighlighters() {
    this.cellHighlighters.clear();
  }

  public void setColumnOrder(List<String> columnOrder) {
    this.columnOrder = columnOrder;
  }

  public void setToolTip(Object closure) {
    try {
      for (int rowInd = 0; rowInd < this.values.size(); rowInd++) {
        List<?> row = this.values.get(rowInd);
        List<String> rowToolTips = new ArrayList<>();
        for (int colInd = 0; colInd < row.size(); colInd++) {
          Object[] params = new Object[]{rowInd, colInd, this};
          rowToolTips.add((String) runClosure(closure, params));
        }
        tooltips.add(rowToolTips);
      }
    } catch (Throwable e) {
      throw new IllegalArgumentException("Can not set tooltip using closure.", e);
    }
  }

  public List<List<String>> getTooltips() {
    return tooltips;
  }

  public static List<Map<String, Object>> getValuesAsRows(List<List<?>> values, List<String> columns) {
    List<Map<String, Object>> rows = new ArrayList<Map<String, Object>>();
    if (columns != null && values != null) {

      for (List<?> value : values) {
        Map<String, Object> m = new HashMap<String, Object>();
        for (int c = 0; c < columns.size(); c++) {
          if (value.size() > c)
            m.put(columns.get(c), value.get(c));
        }
        rows.add(m);
      }
    } else {
      throw new IllegalArgumentException("Method 'getValuesAsRows' doesn't supported for this table");
    }
    return rows;
  }

  public static List<List<?>> getValuesAsMatrix(List<List<?>> values) {
    return values;
  }

  public static Map<String, Object> getValuesAsDictionary(List<List<?>> values) {
    Map<String, Object> m = new HashMap<String, Object>();
    for (List<?> l : values) {
      if (l.size() != 2)
        throw new IllegalArgumentException("Method 'getValuesAsDictionary' doesn't supported for this table");
      ;
      m.put(l.get(0).toString(), l.get(1));
    }
    return m;
  }

  public  List<Map<String, Object>> getValuesAsRows(){
    return getValuesAsRows(values, columns);
  }

  public  List<List<?>> getValuesAsMatrix(){
    return getValuesAsMatrix( values);
  }

  public  Map<String, Object> getValuesAsDictionary(){
    return getValuesAsDictionary(values);
  }

  private Object getValueForSerializer(Object value, BeakerObjectConverter serializer){
    if (value != null) {
      String clazz = serializer.convertType(value.getClass().getName());
      if (BasicObjectSerializer.TYPE_LONG.equals(clazz) ||
          BasicObjectSerializer.TYPE_BIGINT.equals(clazz)){
        return value.toString();
      }
      return value;
    }
    return null;
  }


  public List<List<?>> getValues() { return values; }
  public List<String> getColumnNames() { return columns; }
  public List<String> getTypes() { return classes; }
  public String getSubtype() {
    return subtype;
  }

}

File: beakerx/beakerx/tabledisplay/tabledisplay.py
# Copyright 2017 TWO SIGMA OPEN SOURCE, LLC
#
# Licensed under the Apache License, Version 2.0 (the "License")
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import numpy as np
from beakerx.beakerx_widgets import *
from beakerx.tabledisplay.tableitems import *
from beakerx.utils import *
from ipykernel.comm import Comm
from pandas import DataFrame, RangeIndex, MultiIndex, DatetimeIndex
from traitlets import Unicode, Dict


class Table(BaseObject):
    NAT_VALUE = "NaT"

    def __init__(self, *args, **kwargs):

        self.values = []
        self.types = []
        types_map = dict()
        self.columnNames = []
        self.hasIndex = None
        if isinstance(args[0], DataFrame):
            self.convert_from_pandas(args, types_map)
        elif isinstance(args[0], dict):
            self.convert_from_dict(args)
        elif isinstance(args[0], list):
            self.convert_from_list(args, types_map)

        self.headersVertical = False
        self.headerFontSize = None
        self.contextMenuItems = []
        self.alignmentForType = {}
        self.tooManyRows = False
        self.stringFormatForColumn = {}
        self.subtype = "ListOfMaps"
        self.stringFormatForType = {}
        self.fontColor = []
        self.contextMenuTags = {}
        self.cellHighlighters = []
        self.type = "TableDisplay"
        self.timeZone = None
        self.tooltips = []
        self.columnsFrozen = {}
        self.rendererForType = {}
        self.doubleClickTag = None
        self.alignmentForColumn = {}
        self.columnOrder = []
        self.rendererForColumn = {}
        self.dataFontSize = None
        self.columnsFrozenRight = {}
        self.columnsVisible = {}
        self.hasDoubleClickAction = False
        self.filteredValues = None

    def convert_from_dict(self, args):
        self.columnNames.append("Key")
        self.columnNames.append("Value")
        for key in args[0].keys():
            row = [key, args[0].get(key, "")]
            self.values.append(row)

    def convert_from_list(self, args, types_map):
        for element in args[0]:
            for key in element.keys():
                if key not in self.columnNames:
                    self.columnNames.append(key)
                    column_type = self.convert_type(type(element[key]))
                    self.types.append(column_type)
                    types_map[key] = column_type
                elif types_map[key] != "string":
                    type_for_key = types_map[key]
                    column_type = self.convert_type(type(element[key]))
                    if type_for_key != column_type:
                        self.types[self.columnNames.index(key)] = "string"
                        types_map[key] = "string"
        for element in args[0]:
            row = []
            for columnName in self.columnNames:
                value = element.get(columnName, "")
                value_type = types_map.get(columnName)
                row.append(self.convert_value(value, value_type))
            self.values.append(row)

    def convert_from_pandas(self, args, types_map):
        self.columnNames = args[0].columns.tolist()
        if args[0].index.name is not None and args[0].index.name in self.columnNames:
            self.columnNames.remove(args[0].index.name)

        column = None
        for column in self.columnNames:
            column_type = self.convert_type(args[0].dtypes[column].name)
            self.types.append(column_type)
            types_map[column] = column_type
        for index in range(len(args[0])):
            row = []
            for columnName in self.columnNames:
                value = args[0][columnName].get_values()[index]
                value_type = types_map.get(columnName)
                row.append(self.convert_value(value, value_type))
            if not isinstance(args[0].index, RangeIndex):
                index_type = self.convert_type(args[0].index.dtype)
                index_values = args[0].index.get_values()[index]
                tz = self.get_tz(args[0].index)
                row[:0] = [self.convert_value(index_values, index_type, tz)]
            self.values.append(row)

        if not isinstance(args[0].index, RangeIndex) and column is not None:
            self.hasIndex = "true"
            if isinstance(args[0].index, MultiIndex):
                columns = list(map(lambda x: self.convert_none_to_index_name(x), args[0].index.names))
                self.columnNames[:0] = [', '.join(columns)]
            else:
                self.columnNames[:0] = [args[0].index.name]
            self.types[:0] = [self.convert_type(args[0].index.dtype)]

    @staticmethod
    def get_tz(index):
        if not isinstance(index, DatetimeIndex):
            return None
        tz = index.tz
        if tz is None:
            return None
        return tz.zone

    @staticmethod
    def convert_none_to_index_name(x):
        if x is None:
            return "index"
        else:
            return x

    @staticmethod
    def convert_value(value, value_type, tz=None):
        if value_type == "time":
            if np.isnat(value):
                return str(Table.NAT_VALUE)
            return DateType(value, tz)
        else:
            return value

    @staticmethod
    def convert_type(object_type):
        type_name = str(object_type)
        if "float" in type_name:
            return "double"
        elif "int" in type_name:
            return "integer"
        elif "datetime" in type_name:
            return "time"
        elif "bool" in type_name:
            return "boolean"
        else:
            return "string"

    def setToolTip(self, configTooltip):
        for row_ind in range(0, len(self.values)):
            row = self.values[row_ind]
            rowToolTips = []
            for col_ind in range(0, len(row)):
                rowToolTips.append(configTooltip(row_ind, col_ind, self))
            self.tooltips.append(rowToolTips)

    def setDataFontSize(self, dataFontSize):
        self.dataFontSize = dataFontSize

    def setHeaderFontSize(self, headerFontSize):
        self.headerFontSize = headerFontSize

    def setFontColorProvider(self, colorProvider):
        for row_ind in range(0, len(self.values)):
            row = self.values[row_ind]
            row_font_colors = []
            for col_ind in range(0, len(row)):
                row_font_colors.append(colorProvider(row_ind, col_ind, self))
            self.fontColor.append(row_font_colors)

    def setHeadersVertical(self, headersVertical):
        self.headersVertical = headersVertical

    def setRowFilter(self, filter_row):
        self.filteredValues = []
        for row_ind in range(0, len(self.values)):
            if filter_row(row_ind, self.values):
                self.filteredValues.append(self.values[row_ind])


class TableDisplay(BeakerxDOMWidget):
    _view_name = Unicode('TableDisplayView').tag(sync=True)
    _model_name = Unicode('TableDisplayModel').tag(sync=True)
    _view_module = Unicode('beakerx').tag(sync=True)
    _model_module = Unicode('beakerx').tag(sync=True)
    _model_module_version = Unicode('*').tag(sync=True)
    _view_module_version = Unicode('*').tag(sync=True)

    model = Dict().tag(sync=True)
    contextMenuListeners = dict()

    def __init__(self, *args, **kwargs):
        super(TableDisplay, self).__init__(**kwargs)
        self.chart = Table(*args, **kwargs)
        self.model = self.chart.transform()
        self.on_msg(self.handle_msg)
        self.details = None

    def setAlignmentProviderForType(self, type, alignmentProvider):
        if isinstance(type, ColumnType):
            self.chart.alignmentForType[type.value] = alignmentProvider
            self.model = self.chart.transform()
        return self

    def setAlignmentProviderForColumn(self, column_name, display_alignment):
        if isinstance(display_alignment, TableDisplayAlignmentProvider):
            self.chart.alignmentForColumn[column_name] = display_alignment.value
        self.model = self.chart.transform()
        return self

    def setStringFormatForTimes(self, time_unit):
        self.setStringFormatForType(ColumnType.Time, TableDisplayStringFormat.getTimeFormat(time_unit))
        return self

    def setStringFormatForType(self, type, formater):
        if isinstance(type, ColumnType):
            self.chart.stringFormatForType[type.value] = formater
            self.model = self.chart.transform()
            return self

    def setStringFormatForColumn(self, column, formater):
        self.chart.stringFormatForColumn[column] = formater
        self.model = self.chart.transform()
        return self

    def setRendererForColumn(self, column, renderer):
        self.chart.rendererForColumn[column] = renderer
        self.model = self.chart.transform()
        return self

    def setRendererForType(self, type, renderer):
        if isinstance(type, ColumnType):
            self.chart.rendererForType[type.value] = renderer
            self.model = self.chart.transform()
        return self

    def setColumnFrozen(self, column, visible):
        self.chart.columnsFrozen[column] = visible
        self.model = self.chart.transform()
        return self

    def setColumnFrozenRight(self, column, visible):
        self.chart.columnsFrozenRight[column] = visible
        self.model = self.chart.transform()
        return self

    def setColumnVisible(self, column, visible):
        self.chart.columnsVisible[column] = visible
        self.model = self.chart.transform()
        return self

    def setColumnOrder(self, order):
        self.chart.columnOrder = order
        self.model = self.chart.transform()
        return self

    def removeAllCellHighlighters(self):
        self.chart.cellHighlighters = []
        self.model = self.chart.transform()
        return self

    def addCellHighlighter(self, highlighter):
        if isinstance(highlighter, Highlighter):
            self.chart.cellHighlighters.append(highlighter)
            self.model = self.chart.transform()
        return self

    def setDoubleClickAction(self, listener):
        if listener is not None:
            if isinstance(listener, str):
                self.doubleClickListener = None
                self.chart.doubleClickTag = listener
            elif isinstance(listener, types.FunctionType):
                self.doubleClickListener = listener
                self.chart.doubleClickTag = None
                self.chart.hasDoubleClickAction = True

            self.model = self.chart.transform()

    def setTimeZone(self, timezone):
        self.chart.timeZone = timezone
        self.model = self.chart.transform()
        return self

    def addContextMenuItem(self, name, func):
        self.contextMenuListeners[name] = func
        self.chart.contextMenuItems.append(name)
        self.model = self.chart.transform()

    def doubleClickListener(self, row, column, tabledisplay):
        pass

    def handle_msg(self, tabledisplay, params, list):
        self.details = TableActionDetails(params)
        if params['event'] == 'DOUBLE_CLICK':
            self.doubleClickListener(params['row'], params['column'], tabledisplay)
            self.model = self.chart.transform()
        if params['event'] == 'CONTEXT_MENU_CLICK':
            func = self.contextMenuListeners.get(params['itemKey'])
            if func is not None:
                if isinstance(func, str):
                    self._run_by_tag(func)
                else:
                    func(params['row'], params['column'], tabledisplay)
                    self.model = self.chart.transform()
        if params['event'] == 'actiondetails':
            if params['params']['actionType'] == 'DOUBLE_CLICK':
                self._run_by_tag(self.chart.doubleClickTag)

    def _run_by_tag(self, tag):
        arguments = dict(target_name='beakerx.tag.run')
        comm = Comm(**arguments)
        msg = {'runByTag': tag}
        state = {'state': msg}
        comm.send(data=state, buffers=[])

    def updateCell(self, row, columnName, value):
        row = self.chart.values[row]
        col_index = self.chart.columnNames.index(columnName)
        row[col_index] = value

    def sendModel(self):
        self.model = self.chart.transform()

    @property
    def values(self):
        return self.chart.values

    def setToolTip(self, configTooltip):
        self.chart.setToolTip(configTooltip)
        self.model = self.chart.transform()

    def setDataFontSize(self, dataFontSize):
        self.chart.setDataFontSize(dataFontSize)
        self.model = self.chart.transform()

    def setHeaderFontSize(self, headerFontSize):
        self.chart.setHeaderFontSize(headerFontSize)
        self.model = self.chart.transform()

    def setFontColorProvider(self, colorProvider):
        self.chart.setFontColorProvider(colorProvider)
        self.model = self.chart.transform()

    def setHeadersVertical(self, headersVertical):
        self.chart.setHeadersVertical(headersVertical)
        self.model = self.chart.transform()

    def setRowFilter(self, filter_row):
        self.chart.setRowFilter(filter_row)
        self.model = self.chart.transform()


class TableActionDetails:
    def __init__(self, params):
        if 'params' in params:
            self.row = params['params'].get('row')
            self.col = params['params'].get('col')
            self.contextMenuItem = params['params'].get('contextMenuItem')
            self.actionType = params['params'].get('actionType')
            self.tag = params['params'].get('tag')
        else:
            self.row = params.get('row')
            self.col = params.get('column')
            self.actionType = params.get('event')
            self.contextMenuItem = params.get('itemKey')
            self.tag = params.get('tag')

    def __str__(self):
        return '{} {} {} {}'.format(self.actionType, self.row, self.col, self.tag)
